src/Workspaces/Core/MSBuild/Rpc/RpcClient.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Rpc
+{
+    internal sealed class RpcClient
+    {
+        private readonly TextWriter _sendingStream;
+        private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+        private readonly TextReader _receivingStream;
+
+        private readonly ConcurrentDictionary<int, (TaskCompletionSource<object?>, System.Type? expectedReturnType)> _outstandingRequests = new();
+        private volatile int _nextRequestId = 0;
+
+        public RpcClient(Stream sendingStream, Stream receivingStream)
+        {
+            _sendingStream = new StreamWriter(sendingStream, Encoding.UTF8);
+            _receivingStream = new StreamReader(receivingStream, Encoding.UTF8);
+        }
+
+        public void Start()
+        {
+            Task.Run(async () =>
```

Comment:
fire and forget task in this void method makes me unhappy.

Reply:
This is still a start method, but there's now an equivalent disconnect to stop it.

Reply:
I could make this also return the task, but the caller would just ignore it that case.